pennylane/templates/quantum_quantities/swap_test.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the SwapTest template.
+"""
+
+import pennylane as qml
+import numpy as np
+from pennylane.operation import Operation, AnyWires, Wires
+from pennylane.ops import Hadamard, CSWAP
+from pennylane.measure import custom_process
+
+
+class SwapTest(Operation):
+    r"""A class that implements the SWAPTest.
+
+    The Swap Test is a method used to determine the fidelity between two quantum registers. It is computed
+    using an auxiliary qubit and controlled swap gates between the two registers. The Auxiliary qubit is then
+    repeatedly measured and the outcome of theses measurements is used to determine the fidelity.
+
+    .. math::
+        F(\psi, \phi) = 1 - \frac{2}{n} * \sum_{i=1}^{n} m_{i}
+
+    """
+
+    num_wires = AnyWires
+    grad_recipe = None
+
+    def __init__(self, ancilla, q_reg1, q_reg2, do_queue=True, id=None):
+        wires = [ancilla] + q_reg1 + q_reg2
+
+        if len(q_reg1) != len(q_reg2):
+            raise ValueError(
+                f"The two quantum registers must be the same size to compare them via SWAPTest, "
+                f"got: {q_reg1}, {q_reg2}"
+            )
+
+        self.ac = ancilla
+        self.q_reg1 = q_reg1
+        self.q_reg2 = q_reg2
+
+        super().__init__(wires=wires, do_queue=do_queue, id=id)
+
+    @property
+    def num_params(self):
+        return 0
+
+    def expand(self):
+        with qml.tape.QuantumTape() as tape:
+            len_register = len(self.q_reg1)
+
+            Hadamard(wires=self.ac)  # apply hadamard to ancilla qubit
+            for i in range(len_register):
+                CSWAP(
+                    wires=Wires([self.ac, self.q_reg1[i], self.q_reg2[i]])
+                )  # swap qubit registers
+            Hadamard(wires=self.ac)  # apply hadamard again
+
+        return tape
+
+    def __call__(self, *args, **kwargs):
+        base_measurement = "sample"
+
+        def compute_ip_from_samples(samples):
+            num_samples = len(samples)
+            sum_measurements = np.sum(samples)
+
+            return 1 - (2 / num_samples) * sum_measurements
+
+        return custom_process(compute_ip_from_samples, base_measurement, wires=self.ac)
```

Comment:
I've been pondering this a lot, trying to think from a UI perspective if there is a more intuitive way to combine measurements with operations.

Ideally, it would be great if we could simply do

```python
return qml.SwapTest(*params, wires)
```

but then we run into an issue where validation would fail because the QNode always expects to terminate in a measurement process.

Another solution could be to treat the `MeasurementProcess` as a _mixin_:

```python
class SwapTest(Operation, MeasurementProcess):
```

This is really nice conceptually --- we are creating an object that is both an operation _and_ a measurement process! But it also adds a _lot_ of complexity, and I'm not sure is even possible without a big rewrite of the `MeasurementProcess` class.

A couple of other options I can think of include

- `return qml.SwapTest(*params, wires).measure` (but this is pretty ugly)

- `return qml.swap_test(*params, wires)`. That is, we have a _function_ that internally instantiates a swap test operation, and then returns the custom measurement.